ICE: java.lang.AssertionError: assertion failed: isType called on not-a-Type val <none>, at dotty.tools.dotc.core.Symbols$Symbol.asType(Symbols.scala:167)

## Compiler Version

```
Scala compiler version 3.0.1-RC1-bin-SNAPSHOT-git-4410752 -- Copyright 2002-2021, LAMP/EPFL
```

## System Information

```
Linux 5.4.0-66-generic x86_64 GNU/Linux
```

## Minimized Code

```scala
object F { type T[G[X] <: X, F <: G[F]] }
```

## Output (click arrow to expand)

<details>

```scala
-- Error: mutant.scala:1:31 -----------------------------------------------------------------------------------------------------------------
1 |object F { type T[G[X] <: X, F <: G[F]] }
  |                               ^
  |                               Recursion limit exceeded.
  |                               Maybe there is an illegal cyclic reference?
  |                               If that's not the case, you could also try to increase the stacksize using the -Xss JVM option.
  |                               A recurring operation is (inner to outer):
  |
  |                                 type parameters of  <: G[F]
  |                                 type parameters of F
  |                                 type parameters of F
  |                                 type parameters of G[F]
  |                                 type parameters of  <: G[F]
  |                                 type parameters of F
  |                                 type parameters of F
  |                                 type parameters of G[F]
  |                                 type parameters of  <: G[F]
  |                                 type parameters of F
  |                                 ...
  |
  |                                 type parameters of F
  |                                 type parameters of G[F]
  |                                 type parameters of  <: G[F]
  |                                 type parameters of F
  |                                 type parameters of F
  |                                 type parameters of G[F]
  |                                 type parameters of  <: G[F]
  |                                 type parameters of F
  |                                 type parameters of G[F]
  |                                 type parameters of  <: G[F]
-- Error: mutant.scala:1:29 -----------------------------------------------------------------------------------------------------------------
1 |object F { type T[G[X] <: X, F <: G[F]] }
  |                             ^
  |                             Recursion limit exceeded.
  |                             Maybe there is an illegal cyclic reference?
  |                             If that's not the case, you could also try to increase the stacksize using the -Xss JVM option.
  |                             A recurring operation is (inner to outer):
  |
  |                               type parameters of F
  |                               type parameters of F
  |                               type parameters of G[F]
  |                               type parameters of  <: G[F]
  |                               type parameters of F
  |                               type parameters of F
  |                               type parameters of G[F]
  |                               type parameters of  <: G[F]
  |                               type parameters of F
  |                               type parameters of F
  |                               ...
  |
  |                               type parameters of F
  |                               type parameters of F
  |                               type parameters of G[F]
  |                               type parameters of  <: G[F]
  |                               type parameters of F
  |                               type parameters of F
  |                               type parameters of G[F]
  |                               type parameters of  <: G[F]
  |                               type parameters of F
  |                               type parameters of G[F]
exception occurred while typechecking mutant.scala
exception occurred while compiling mutant.scala
java.lang.AssertionError: assertion failed: isType called on not-a-Type val <none> while compiling mutant.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: isType called on not-a-Type val <none>
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.Symbols$Symbol.asType(Symbols.scala:167)
	at dotty.tools.dotc.typer.TypeAssigner.assignType$$anonfun$6(TypeAssigner.scala:461)
	at scala.collection.immutable.List.map(List.scala:250)
	at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:461)
	at dotty.tools.dotc.typer.TypeAssigner.assignType$(TypeAssigner.scala:19)
	at dotty.tools.dotc.typer.Typer.assignType(Typer.scala:106)
	at dotty.tools.dotc.typer.Typer.typeIndexedLambdaTypeTree(Typer.scala:1934)
	at dotty.tools.dotc.typer.Typer.typedTypeDef(Typer.scala:2221)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2675)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2677)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2747)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2813)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2817)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2839)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2889)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2344)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2673)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2677)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2747)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2813)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2817)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2839)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2889)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2467)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2718)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2748)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2813)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2817)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2933)
	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
	at dotty.tools.dotc.Run.compile(Run.scala:150)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.dotc.Driver.process(Driver.scala:199)
	at dotty.tools.dotc.Driver.process(Driver.scala:167)
	at dotty.tools.dotc.Driver.process(Driver.scala:179)
	at dotty.tools.dotc.Driver.main(Driver.scala:209)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala
+++ b/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala
@@ -480,7 +480,12 @@ trait TypeAssigner {
   }
 
   def assignType(tree: untpd.LambdaTypeTree, tparamDefs: List[TypeDef], body: Tree)(using Context): LambdaTypeTree =
-    tree.withType(HKTypeLambda.fromParams(tparamDefs.map(_.symbol.asType), body.tpe))
+    val validParams = tparamDefs.filterConserve { tdef =>
+      val ok = tdef.symbol.isType
+      if !ok then assert(ctx.reporter.errorsReported)
+      ok
+    }
+    tree.withType(HKTypeLambda.fromParams(validParams.map(_.symbol.asType), body.tpe))
 
   def assignType(tree: untpd.MatchTypeTree, bound: Tree, scrutinee: Tree, cases: List[CaseDef])(using Context): MatchTypeTree = {
     val boundType = if (bound.isEmpty) defn.AnyType else bound.tpe

--- a/tests/neg/i12456.scala
+++ b/tests/neg/i12456.scala
@@ -1,0 +1,1 @@
+object F { type T[G[X] <: X, F <: G[F]] } // error // error

